File: core/src/com/nopalsoft/zombiewars/game/WorldGame.java
package com.nopalsoft.zombiewars.game;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.maps.tiled.TiledMapTileLayer;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.Contact;
import com.badlogic.gdx.physics.box2d.ContactImpulse;
import com.badlogic.gdx.physics.box2d.ContactListener;
import com.badlogic.gdx.physics.box2d.Fixture;
import com.badlogic.gdx.physics.box2d.Manifold;
import com.badlogic.gdx.physics.box2d.World;
import com.badlogic.gdx.utils.Array;
import com.nopalsoft.zombiewars.Assets;
import com.nopalsoft.zombiewars.Settings;
import com.nopalsoft.zombiewars.objects.Bullet;
import com.nopalsoft.zombiewars.objects.Personajes;

public class WorldGame {
    static final int STATE_RUNNING = 0;

    final float TIME_TO_SPAWN_ZOMBIE_FRANK;
    final float TIME_TO_SPAWN_ZOMBIE_CUASY;
    final float TIME_TO_SPAWN_ZOMBIE_KID;
    final float TIME_TO_SPAWN_ZOMBIE_MUMMY;
    final float TIME_TO_SPAWN_ZOMBIE_PAN;
    public int state;
    public int tiledWidth;
    public int tiledHeight;
    public World oWorldBox;
    public ObjectCreatorManagerBox2d objectCreatorManager;
    public Vector2 posCamara;
    float timeToSpwanZombieFrank;
    float timeToSpwanZombieCuasy;
    float timeToSpwanZombieKid;
    float timeToSpwanZombieMummy;
    float timeToSpwanZombiePan;
    /**
     * Mis tiles son de 32px, asi que la unidad seria 1/32 con una camara ortograpicha de 10x15 para ver 10 tiles de ancho y 15 de alto. El probema es que mi camara es de 8x4.8f por eso tengo que
     * cambiar la escala, con 1/32 solo veria 8 tiles a lo ancho y de altura 4.8 por como esta configurada la camara.
     * <p>
     * con 1/96 veo 24 tiles a lo ancho
     */
    float unitScale = 1 / 76f;
    float xMin, xMax, yMin;
    Array<Personajes> arrFacingRight;
    Array<Personajes> arrFacingLeft;
    Array<Bullet> arrBullets;
    Array<Body> arrBodies;

    public WorldGame(int nivel) {
        oWorldBox = new World(new Vector2(0, -9.8f), true);
        oWorldBox.setContactListener(new com.nopalsoft.zombiewars.game.WorldGame.Colisiones());

        arrFacingRight = new Array<>();
        arrFacingLeft = new Array<>();
        arrBullets = new Array<>();
        arrBodies = new Array<>();

        objectCreatorManager = new ObjectCreatorManagerBox2d(this);
        new TiledMapManagerBox2d(this, unitScale).createObjetosDesdeTiled(Assets.map);
        tiledWidth = ((TiledMapTileLayer) Assets.map.getLayers().get("1")).getWidth();
        tiledHeight = ((TiledMapTileLayer) Assets.map.getLayers().get("1")).getHeight();

        if (tiledWidth * tiledHeight > 2500) {
            Gdx.app.log("Advertencia de rendimiento", "Hay mas de 2500 tiles " + tiledWidth + " x " + tiledHeight + " = "
                    + (tiledWidth * tiledHeight));
        }

        Gdx.app.log("Tile Width", tiledWidth + "");
        Gdx.app.log("Tile Height", tiledHeight + "");

        xMin = 4.0f;// Inicia en 4 porque la camara esta centrada no en el origen
        xMax = unitScale * tiledWidth * 32 - 4;// Menos 4 porque la camara esta centrada en el origen
        yMin = 2.4f;

        posCamara = new Vector2(xMin, yMin);
        state = STATE_RUNNING;

        if (nivel == 0) {
            TIME_TO_SPAWN_ZOMBIE_KID = 3f;
            TIME_TO_SPAWN_ZOMBIE_CUASY = 10f;
            TIME_TO_SPAWN_ZOMBIE_MUMMY = 15f;
            TIME_TO_SPAWN_ZOMBIE_PAN = 20f;
            TIME_TO_SPAWN_ZOMBIE_FRANK = 25f;
        } else {
            TIME_TO_SPAWN_ZOMBIE_KID = 0f;
            TIME_TO_SPAWN_ZOMBIE_CUASY = 0f;
            TIME_TO_SPAWN_ZOMBIE_MUMMY = 0f;
            TIME_TO_SPAWN_ZOMBIE_PAN = 0f;
            TIME_TO_SPAWN_ZOMBIE_FRANK = 0f;
        }

    }

    public void update(float delta, float accelCamX) {
        oWorldBox.step(delta, 8, 4);
        updateCamara(delta, accelCamX);

        eliminarObjetos();

        spawnStuff(delta);

        oWorldBox.getBodies(arrBodies);

        for (Body body : arrBodies) {
            if (body.getUserData() instanceof Personajes) {
                Personajes obj = (Personajes) body.getUserData();
                if (obj.isFacingLeft)
                    updateFacingLeft(delta, obj);
                else
                    updateFacingRight(delta, obj);
            } else if (body.getUserData() instanceof Bullet) {
                updateBullet(delta, body);
            }
        }
    }

    public void atackaLL() {

        for (Personajes obj : arrFacingLeft) {

            if (obj.attack())

                objectCreatorManager.createBullet(obj);
        }
    }

    public void dieALl() {

        for (Personajes obj : arrFacingLeft) {

            obj.die();
        }
    }

    private void spawnStuff(float delta) {

        timeToSpwanZombieKid += delta;
        if (timeToSpwanZombieKid >= TIME_TO_SPAWN_ZOMBIE_KID) {
            timeToSpwanZombieKid -= TIME_TO_SPAWN_ZOMBIE_KID;
            objectCreatorManager.createZombieKid();
        }

        timeToSpwanZombieCuasy += delta;
        if (timeToSpwanZombieCuasy >= TIME_TO_SPAWN_ZOMBIE_CUASY) {
            timeToSpwanZombieCuasy -= TIME_TO_SPAWN_ZOMBIE_CUASY;
            objectCreatorManager.createZombieCuasy();
        }

        timeToSpwanZombieMummy += delta;
        if (timeToSpwanZombieMummy >= TIME_TO_SPAWN_ZOMBIE_MUMMY) {
            timeToSpwanZombieMummy -= TIME_TO_SPAWN_ZOMBIE_MUMMY;
            objectCreatorManager.createZombieMummy();
        }

        timeToSpwanZombiePan += delta;
        if (timeToSpwanZombiePan >= TIME_TO_SPAWN_ZOMBIE_PAN) {
            timeToSpwanZombiePan -= TIME_TO_SPAWN_ZOMBIE_PAN;
            objectCreatorManager.createZombiePan();
        }

        timeToSpwanZombieFrank += delta;
        if (timeToSpwanZombieFrank >= TIME_TO_SPAWN_ZOMBIE_FRANK) {
            timeToSpwanZombieFrank -= TIME_TO_SPAWN_ZOMBIE_FRANK;
            objectCreatorManager.createZombieFrank();
        }

    }

    private void updateCamara(float delta, float accelCamX) {

        if (accelCamX != 0)
            posCamara.x += (delta * accelCamX);

        if (posCamara.x < xMin * Settings.zoom) {
            posCamara.x = xMin * Settings.zoom;
        } else if (posCamara.x > (xMax - (xMin * (Settings.zoom - 1)))) {
            posCamara.x = xMax - (xMin * (Settings.zoom - 1));
        }

        posCamara.y = yMin * Settings.zoom;

    }

    private void updateFacingRight(float delta, Personajes obj) {
        obj.update(delta);

        int len = arrFacingLeft.size;
        for (int i = 0; i < len; i++) {
            Personajes objMalo = arrFacingLeft.get(i);

            if (obj.position.dst(objMalo.position.x, objMalo.position.y) <= obj.DISTANCE_ATTACK) {
                if (obj.attack())
                    objectCreatorManager.createBullet(obj);

            }
        }
    }

    private void updateFacingLeft(float delta, Personajes obj) {
        obj.update(delta);

        int len = arrFacingRight.size;
        for (int i = 0; i < len; i++) {
            Personajes objBueno = arrFacingRight.get(i);
            if (obj.position.dst(objBueno.position.x, objBueno.position.y) <= obj.DISTANCE_ATTACK) {
                if (obj.attack())
                    objectCreatorManager.createBullet(obj);

            }
        }
    }

    private void updateBullet(float delta, Body body) {
        Bullet obj = (Bullet) body.getUserData();
        obj.update(delta);

        if (obj.position.x > xMax + 3 || obj.position.x < xMin - 3)
            obj.state = Bullet.STATE_DESTROY;

    }

    private void eliminarObjetos() {
        oWorldBox.getBodies(arrBodies);

        for (Body body : arrBodies) {
            if (!oWorldBox.isLocked()) {
                if (body.getUserData() instanceof Personajes) {
                    Personajes obj = (Personajes) body.getUserData();
                    if (obj.state == Personajes.STATE_DEAD && obj.stateTime >= obj.DURATION_DEAD) {
                        if (obj.isFacingLeft)
                            arrFacingLeft.removeValue(obj, true);
                        else
                            arrFacingRight.removeValue(obj, true);

                        oWorldBox.destroyBody(body);
                    }
                } else if (body.getUserData() instanceof Bullet) {
                    Bullet obj = (Bullet) body.getUserData();
                    if (obj.state == Bullet.STATE_DESTROY) {
                        arrBullets.removeValue(obj, true);
                        oWorldBox.destroyBody(body);
                    }
                }

            }
        }
    }

    static class Colisiones implements ContactListener {

        @Override
        public void beginContact(Contact contact) {
            Fixture a = contact.getFixtureA();
            Fixture b = contact.getFixtureB();
            if (a.getBody().getUserData() instanceof Bullet)
                beginContactBulletOtraCosa(a, b);
            else if (b.getBody().getUserData() instanceof Bullet)
                beginContactBulletOtraCosa(b, a);

        }

        private void beginContactBulletOtraCosa(Fixture fixBullet, Fixture otraCosa) {
            Object oOtraCosa = otraCosa.getBody().getUserData();
            Bullet oBullet = (Bullet) fixBullet.getBody().getUserData();

            if (oOtraCosa instanceof Personajes) {
                if (oBullet.state == Bullet.STATE_NORMAL || oBullet.state == Bullet.STATE_MUZZLE) {
                    Personajes obj = (Personajes) oOtraCosa;

                    if (obj.isFacingLeft == oBullet.isFacingLeft)// Si van hacia el mismo lado son amigos
                        return;

                    if (obj.state != Personajes.STATE_DEAD) {

                        obj.getHurt(oBullet.DAMAGE);
                        oBullet.hit();

                    }
                }
            }

        }

        @Override
        public void endContact(Contact contact) {
        }

        @Override
        public void preSolve(Contact contact, Manifold oldManifold) {
        }

        @Override
        public void postSolve(Contact contact, ContactImpulse impulse) {
        }

    }

}

File: core/src/com/nopalsoft/zombiewars/game/WorldGame.kt
package com.nopalsoft.zombiewars.game

import com.badlogic.gdx.utils.Array as GdxArray
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.maps.tiled.TiledMapTileLayer
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.Body
import com.badlogic.gdx.physics.box2d.Contact
import com.badlogic.gdx.physics.box2d.ContactImpulse
import com.badlogic.gdx.physics.box2d.ContactListener
import com.badlogic.gdx.physics.box2d.Fixture
import com.badlogic.gdx.physics.box2d.Manifold
import com.badlogic.gdx.physics.box2d.World
import com.nopalsoft.zombiewars.Assets
import com.nopalsoft.zombiewars.Settings
import com.nopalsoft.zombiewars.objects.Bullet
import com.nopalsoft.zombiewars.objects.GameCharacter

class WorldGame(level: Int) {

    private val state = STATE_RUNNING
    private val tiledWidth = (Assets.map!!.layers["1"] as TiledMapTileLayer).width
    private val tiledHeight = (Assets.map!!.layers["1"] as TiledMapTileLayer).height
    val oWorldBox = World(Vector2(0f, -9.8f), true)
    val objectCreatorManager = ObjectCreatorManagerBox2d(this)
    private val xMin = 4.0F // Starts at 4 because the camera is centered not at the origin
    private val yMin = 2.4F
    val posCamara = Vector2(xMin, yMin)

    private var timeToSpawnZombieFrank = 0F
    private var timeToSpawnZombieCuasy = 0F
    private var timeToSpawnZombieKid = 0F
    private var timeToSpawnZombieMummy = 0F
    private var timeToSpawnZombiePan = 0F

    /*
     * My tiles are 32px, so the unit would be 1/32 with a 10x15 orthographic camera to see 10 tiles wide and 15 high. The problem is that my camera is 8x4.8f so I have
     * to change the scale, with 1/32 I would only see 8 tiles wide and 4.8 high because of how the camera is configured.
     * <p>
     * with 1/96 I see 24 tiles wide
     */
    val unitScale = 1 / 76F

    private val xMax = unitScale * tiledWidth * 32 - 4 // Minus 4 because the camera is centered at the origin

    val arrFacingRight = GdxArray<GameCharacter>()
    val arrFacingLeft = GdxArray<GameCharacter>()
    val arrBullets = GdxArray<Bullet>()
    private val arrBodies = GdxArray<Body>()

    init {
        oWorldBox.setContactListener(Collision())
        TiledMapManagerBox2d(this, unitScale).createTiledObjects(Assets.map!!)

        if (tiledWidth * tiledHeight > 2_500)
            Gdx.app.log("Advertencia de rendimiento", "Hay mas de 2500 tiles $tiledWidth x $tiledHeight = ${tiledWidth * tiledHeight}")

        Gdx.app.log("Tile Width", tiledWidth.toString())
        Gdx.app.log("Tile Height", tiledHeight.toString())

        if (level == 0) {
            TIME_TO_SPAWN_ZOMBIE_KID = 3f
            TIME_TO_SPAWN_ZOMBIE_CUASY = 10f
            TIME_TO_SPAWN_ZOMBIE_MUMMY = 15f
            TIME_TO_SPAWN_ZOMBIE_PAN = 20f
            TIME_TO_SPAWN_ZOMBIE_FRANK = 25f
        } else {
            TIME_TO_SPAWN_ZOMBIE_KID = 0f
            TIME_TO_SPAWN_ZOMBIE_CUASY = 0f
            TIME_TO_SPAWN_ZOMBIE_MUMMY = 0f
            TIME_TO_SPAWN_ZOMBIE_PAN = 0f
            TIME_TO_SPAWN_ZOMBIE_FRANK = 0f
        }
    }

    fun update(delta: Float, accelCamX: Float) {
        oWorldBox.step(delta, 8, 4)
        updateCamara(delta, accelCamX)

        deleteObjects()

        spawnStuff(delta)

        oWorldBox.getBodies(arrBodies)

        for (body in arrBodies) {
            if (body.userData is GameCharacter) {
                val obj = body.userData as GameCharacter
                if (obj.isFacingLeft) updateFacingLeft(delta, obj)
                else updateFacingRight(delta, obj)
            } else if (body.userData is Bullet) {
                updateBullet(delta, body)
            }
        }
    }

    fun attackAll() {
        for (obj in arrFacingLeft) {
            if (obj.attack()) objectCreatorManager.createBullet(obj)
        }
    }

    fun dieAll() {
        for (obj in arrFacingLeft) {
            obj.die()
        }
    }

    private fun spawnStuff(delta: Float) {
        timeToSpawnZombieKid += delta
        if (timeToSpawnZombieKid >= TIME_TO_SPAWN_ZOMBIE_KID) {
            timeToSpawnZombieKid -= TIME_TO_SPAWN_ZOMBIE_KID
            objectCreatorManager.createZombieKid()
        }

        timeToSpawnZombieCuasy += delta
        if (timeToSpawnZombieCuasy >= TIME_TO_SPAWN_ZOMBIE_CUASY) {
            timeToSpawnZombieCuasy -= TIME_TO_SPAWN_ZOMBIE_CUASY
            objectCreatorManager.createZombieCuasy()
        }

        timeToSpawnZombieMummy += delta
        if (timeToSpawnZombieMummy >= TIME_TO_SPAWN_ZOMBIE_MUMMY) {
            timeToSpawnZombieMummy -= TIME_TO_SPAWN_ZOMBIE_MUMMY
            objectCreatorManager.createZombieMummy()
        }

        timeToSpawnZombiePan += delta
        if (timeToSpawnZombiePan >= TIME_TO_SPAWN_ZOMBIE_PAN) {
            timeToSpawnZombiePan -= TIME_TO_SPAWN_ZOMBIE_PAN
            objectCreatorManager.createZombiePan()
        }

        timeToSpawnZombieFrank += delta
        if (timeToSpawnZombieFrank >= TIME_TO_SPAWN_ZOMBIE_FRANK) {
            timeToSpawnZombieFrank -= TIME_TO_SPAWN_ZOMBIE_FRANK
            objectCreatorManager.createZombieFrank()
        }
    }

    private fun updateCamara(delta: Float, accelCamX: Float) {
        if (accelCamX != 0f) posCamara.x += (delta * accelCamX)

        if (posCamara.x < xMin * Settings.zoom) {
            posCamara.x = xMin * Settings.zoom
        } else if (posCamara.x > (xMax - (xMin * (Settings.zoom - 1)))) {
            posCamara.x = xMax - (xMin * (Settings.zoom - 1))
        }

        posCamara.y = yMin * Settings.zoom
    }

    private fun updateFacingRight(delta: Float, obj: GameCharacter) {
        obj.update(delta)

        val len = arrFacingLeft.size
        for (i in 0 until len) {
            val objMalo = arrFacingLeft[i]

            if (obj.position.dst(objMalo.position.x, objMalo.position.y) <= obj.attackDistance) {
                if (obj.attack()) objectCreatorManager.createBullet(obj)
            }
        }
    }

    private fun updateFacingLeft(delta: Float, obj: GameCharacter) {
        obj.update(delta)

        val len = arrFacingRight.size
        for (i in 0 until len) {
            val objBueno = arrFacingRight[i]
            if (obj.position.dst(objBueno.position.x, objBueno.position.y) <= obj.attackDistance) {
                if (obj.attack()) objectCreatorManager.createBullet(obj)
            }
        }
    }

    private fun updateBullet(delta: Float, body: Body) {
        val obj = body.userData as Bullet
        obj.update(delta)

        if (obj.position.x > xMax + 3 || obj.position.x < xMin - 3) obj.state = Bullet.STATE_DESTROY
    }

    private fun deleteObjects() {
        oWorldBox.getBodies(arrBodies)

        for (body in arrBodies) {
            if (!oWorldBox.isLocked) {
                if (body.userData is GameCharacter) {
                    val obj = body.userData as GameCharacter
                    if (obj.state == GameCharacter.STATE_DEAD && obj.stateTime >= obj.deathAnimationDuration) {
                        if (obj.isFacingLeft) arrFacingLeft.removeValue(obj, true)
                        else arrFacingRight.removeValue(obj, true)

                        oWorldBox.destroyBody(body)
                    }
                } else if (body.userData is Bullet) {
                    val obj = body.userData as Bullet
                    if (obj.state == Bullet.STATE_DESTROY) {
                        arrBullets.removeValue(obj, true)
                        oWorldBox.destroyBody(body)
                    }
                }
            }
        }
    }

    companion object {

        private const val STATE_RUNNING = 0
        private var TIME_TO_SPAWN_ZOMBIE_FRANK = 0f
        private var TIME_TO_SPAWN_ZOMBIE_CUASY = 0f
        private var TIME_TO_SPAWN_ZOMBIE_KID = 0f
        private var TIME_TO_SPAWN_ZOMBIE_MUMMY = 0f
        private var TIME_TO_SPAWN_ZOMBIE_PAN = 0f
    }


    class Collision : ContactListener {
        override fun beginContact(contact: Contact) {
            val a = contact.fixtureA
            val b = contact.fixtureB
            if (a.body.userData is Bullet) beginContactBulletOtraCosa(a, b)
            else if (b.body.userData is Bullet) beginContactBulletOtraCosa(b, a)
        }

        private fun beginContactBulletOtraCosa(fixBullet: Fixture, otraCosa: Fixture) {
            val oOtraCosa = otraCosa.body.userData
            val oBullet = fixBullet.body.userData as Bullet

            if (oOtraCosa is GameCharacter) {
                if (oBullet.state == Bullet.STATE_NORMAL || oBullet.state == Bullet.STATE_MUZZLE) {
                    val obj = oOtraCosa

                    if (obj.isFacingLeft == oBullet.isFacingLeft)  // Si van hacia el mismo lado son amigos
                        return

                    if (obj.state != GameCharacter.STATE_DEAD) {
                        obj.getHurt(oBullet.DAMAGE)
                        oBullet.hit()
                    }
                }
            }
        }

        override fun endContact(contact: Contact) {
        }

        override fun preSolve(contact: Contact, oldManifold: Manifold) {
        }

        override fun postSolve(contact: Contact, impulse: ContactImpulse) {
        }
    }
}